qutip/hsolve.py

```diff
@@ -0,0 +1,310 @@
+        The coupling between system and bath.
+    gam: Float
+        Bath cutoff frequency.
+    lam0: Float
+        Coupling strength.
+    Nc: Integer
+        Cutoff parameter.
+    N: Integer
+        Number of matsubara terms.
+    w_th: Float
+        Temperature. 
+    options : :class:`qutip.Options`
+        With options for the solver.
+     
+    Returns
+    -------
+    output: Result
+        System evolution.
+    """
+    if options is None:
+        options = Options()
+    
+    # Set up terms of the matsubara and tanimura boundaries
+
+    # Parameters and hamiltonian
+    hbar=1.
+    kb=1.
+
+    # Set by system 
+    dimensions = dims(H)
+    Nsup = dimensions[0][0] * dimensions[0][0] 
+    unit = qeye(dimensions[0])
+
+    # Ntot is the total number of ancillary elements in the hierarchy
+    Ntot = int(round(factorial(Nc+N) / (factorial(Nc) * factorial(N))))
+    c0 =(lam0 * gam * (cot(gam * hbar / (2. * kb * w_th)) - (1j))) / hbar
+    LD1 = -2.* spre(Q) * spost(Q.dag()) + spre(Q.dag() * Q) + spost(Q.dag() * Q)
+    pref = ((2. * lam0 * kb * w_th / (gam * hbar)) -1j * lam0) / hbar
+    gj = 2 * np.pi * kb * w_th / hbar
+    L12 = -pref * LD1 + (c0 / gam) * LD1 
+
+    for i1 in range(1,N):
+        ci = (4 * lam0 * gam * kb * w_th * i1
+        	  * gj/((i1 * gj)**2 - gam**2)) / (hbar**2)
+        L12 = L12 + (ci / gj) * LD1
+
+    # Setup liouvillian
+
+    L = liouvillian(H, [L12])
+    Ltot = L.data
+    unit = sp.csr_matrix(np.identity(Ntot))
+    Lbig = sp.kron(unit,Ltot.tocsr())
+    rho0big1 = np.zeros((Nsup * Ntot), dtype=complex)
+
+
+    # Prepare initial state:
+
+    rhotemp =  mat2vec(np.array(psi0.full(), dtype=complex))
+
+    for idx,element in enumerate(rhotemp):
+        rho0big1[idx] = element[0]
+    rho0big = sp.csr_matrix(rho0big1)
+    
+    nstates, state2idx, idx2state = enr_state_dictionaries([Nc+1]*(N),Nc)
+    for nlabelt in state_number_enumerate([Nc+1]*(N),Nc):
+        nlabel = list(nlabelt)                    
+        ntotalcheck = 0
+        for ncheck in range(N):
+            ntotalcheck = ntotalcheck + nlabel[ncheck]                            
+        current_pos = int(round(state2idx[tuple(nlabel)]))
+        Ltemp = sp.lil_matrix((Ntot, Ntot))
+        Ltemp[current_pos,current_pos] = 1
+        Ltemp.tocsr()
+        Lbig = Lbig + sp.kron(Ltemp,(-nlabel[0] * gam * spre(unit).data))
+        
+        for kcount in range(1,N):
+            Lbig = Lbig + sp.kron(Ltemp,(-nlabel[kcount] * (kcount * gj)
+            	   			 * spre(unit).data))
+        
+        for kcount in range(N):
+            if nlabel[kcount] >= 1:
+            # find the position of the neighbour
+                nlabeltemp = copy(nlabel)
+                nlabel[kcount] = nlabel[kcount] -1
+                current_pos2 = int(round(state2idx[tuple(nlabel)]))
+                Ltemp = sp.lil_matrix(np.zeros((Ntot,Ntot)))
+                Ltemp[current_pos, current_pos2] = 1
+                Ltemp.tocsr()
```

Comment:
In general COO is prefered over LIL since conversion to and from CSR and CSC is faster